pkg/controller/extensions/extension/extension.go

```diff
@@ -0,0 +1,251 @@
+func Add(mgr manager.Manager) error {
+	r := &Reconciler{
+		kube:    mgr.GetClient(),
+		factory: &extensionHandlerFactory{},
+	}
+
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Extension
+	err = c.Watch(&source.Kind{Type: &v1alpha1.Extension{}}, &controllerHandler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile reads that state of the Extension for a Instance object and makes changes based on the state read
+// and what is in the Instance.Spec
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("reconciling", "kind", v1alpha1.ExtensionKindAPIVersion, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	// fetch the CRD instance
+	i := &v1alpha1.Extension{}
+	if err := r.kube.Get(ctx, req.NamespacedName, i); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	handler := r.factory.newHandler(ctx, i, r.kube)
+
+	return handler.sync(ctx)
+}
+
+type handler interface {
+	sync(context.Context) (reconcile.Result, error)
+	create(context.Context) (reconcile.Result, error)
+	update(context.Context) (reconcile.Result, error)
+}
+
+type extensionHandler struct {
+	kube client.Client
+	i    *v1alpha1.Extension
+}
+
+type factory interface {
+	newHandler(context.Context, *v1alpha1.Extension, client.Client) handler
+}
+
+type extensionHandlerFactory struct{}
+
+func (f *extensionHandlerFactory) newHandler(ctx context.Context, i *v1alpha1.Extension, kube client.Client) handler {
+	return &extensionHandler{
+		kube: kube,
+		i:    i,
+	}
+}
+
+// ************************************************************************************************
+// Syncing/Creating functions
+// ************************************************************************************************
+func (h *extensionHandler) sync(ctx context.Context) (reconcile.Result, error) {
+	if h.i.Status.ControllerRef == nil {
+		return h.create(ctx)
+	}
+
+	return h.update(ctx)
+}
+
+func (h *extensionHandler) create(ctx context.Context) (reconcile.Result, error) {
+	// create RBAC permissions
+	if err := h.processRBAC(ctx); err != nil {
+		return fail(ctx, h.kube, h.i, reasonCreatingRBAC, err.Error())
+	}
+
+	// create controller deployment
+	if err := h.processDeployment(ctx); err != nil {
+		return fail(ctx, h.kube, h.i, reasonCreatingDeployment, err.Error())
+	}
+
+	// the extension has successfully been created, the extension is ready
+	h.i.Status.UnsetAllConditions()
+	h.i.Status.SetReady()
+	return requeueOnSuccess, h.kube.Status().Update(ctx, h.i)
+}
+
+func (h *extensionHandler) update(ctx context.Context) (reconcile.Result, error) {
+	return reconcile.Result{}, nil
```

Comment:
I think this is the perfect place to write a log message or better yet - Event with information that "Update" is not supported. Ideally, this should be handled in the webhook.